### ISSUE TITLE
Gateway not connecting to updated orderer address

### ISSUE THREAD

**Issue by sapthasurendran** [hyperledger/fabric-gateway#318]
Network Details:
3 org 2 peer each  ,5 orderers
Gateway peer:peer.org1.example.com

Steps:

- Started the network 
- started the node client app
- Stopped orderer2.example.com:7051 container
- Made successful config update transaction to the live network to update the orderer2 address port and consenter port  from 
7051 to 7057. Fetched latest block to confirm the update has been successful. The config block had updated values for orderer2. 
- Restarted orderer2.example.com with new port 7057

Issue: 
Observed that the gateway peer was still trying to connect to old orderer at 7051, which was failing with SERVICE_UNAVAILABLE error.
The client app doesn't throw any errors and transactions are successful

Gateway Peer Log : https://github.com/hyperledger/fabric-gateway/files/7554781/gatewayPeer.log
Orderer2. log : https://github.com/hyperledger/fabric-gateway/files/7554794/orderer2.log

**Comment by sapthasurendran** [comment]
@andrew-coleman  This could be an issue at the orderer end. 
The leader order still tried to communicate  on old port.
After restarting the other orderers, connections with established on the new port. 
I will update once this gets resolved.

**Comment by sapthasurendran** [comment]
@andrew-coleman  @denyeart 
Retested the scenario, I could resolve the orderer side issues with the help from Param. I was missing on system channel updates.
Orderer is now taking part in the block creation but the gateway is not connecting to the new address.

Attaching logs here:
[orderer2.log](https://github.com/hyperledger/fabric-gateway/files/7625648/orderer2.log)
[gatewaylog.log](https://github.com/hyperledger/fabric-gateway/files/7625646/gatewaylog.log)
[orderer1.log](https://github.com/hyperledger/fabric-gateway/files/7625647/orderer1.log)

**Comment by denyeart** [comment]
@andrew-coleman It appears this problem is related to the other problem I mentioned where we get an error message in gateway upon processing block 0 config:

`2021-12-04 08:10:50.526 EST 01f0 ERRO [gateway] configUpdate -> Failed update orderer config channel=mychannel err="failed to get config for channel [mychannel]: could not get last config for channel mychannel"`

The problem is that config event callback registry.configUpdate() is called BEFORE the config transaction is validated or committed. So we get the orderers from the LAST config, not the NEW config. At block 0 there is no last config so we get the error above. I think you need to get the orderer endpoints from the new config bundle that gets passed to configUpdate(), rather than looking for the last config in discovery service (I believe discovery service gets updated in parallel via the same config event mechanism).

The only reason it somewhat works in the existing tests, is because after the genesis block there is usually another config event for updating anchor peers, and when processing this config event, gateway gets the orderer endpoints from the last config.

In addition to this issue in gateway, I think there is a broader Fabric issue in that the config event callbacks happen before the config transaction gets validated. If the config transaction ends up getting invalidated, the config event should not fire.

**Comment by denyeart** [comment]
@sapthasurendran @davidkel Fix is merged into main and release-2.4 branches. Could you re-test? If all looks good we'll get v2.4.1 sooner than later.

**Comment by sapthasurendran** [comment]
@andrew-coleman @denyeart 
Tested with latest gateway version, still I find the gateway connecting to the old port. 
[gateway.log](https://github.com/hyperledger/fabric-gateway/files/7692398/gateway.log)

**Comment by andrew-coleman** [comment]
The attached log shows that you are running an old version of the peer that doesn't contain this fix:
 Commit SHA: 516bc8b

**Comment by sapthasurendran** [comment]
@andrew-coleman @denyeart @davidkel 

The latest peer fixes the issue. 

The gateway is connecting to updated orderer port , can confirm from the below log


```
2021-12-10 14:47:52.640 UTC 0852 INFO [gateway] Submit -> Sending transaction to orderer txID=1f66c10c32317da3af5c04824b47ea909f462ec89c8cd879734bd987e58dc5fb endpoint=orderer2.example.com:7057
2021-12-10 14:47:52.652 UTC 0853 INFO [comm.grpc.server] 1 -> unary call completed grpc.service=gateway.Gateway grpc.method=Submit grpc.request_deadline=2021-12-10T14:48:02.64Z grpc.peer_address=172.18.0.1:65160 grpc.code=OK grpc.call_duration=11.7117ms
```

> The attached log shows that you are running an old version of the peer that doesn't contain this fix: Commit SHA: 516bc8b

Yes I was testing with older version.
